raft-tech/TANF-app
# Buildpacks CI/CD Implementation

### Summary of Changes
Closes #844 

* Refactors the Circle CI config.yml to utilize re-usable commands for many common deployment steps
* Added Circle CI `job` configuration for deployment environments of dev, staging and prod
* Utilizes new buildpack deployment scripts instead of previous docker based deployment script
* Updates GitHub Action for label based deployment to call the new `deploy-dev` job
* Updates docker image used for running deployment commands to a non-legacy CircleCI image

### How to Test
* Delete the existing `tdp-backend` and `tdp-frontend` apps from the Cloud.gov Dashboard - Docker based app deployments cannot be converted to buildpack deployments as part of a rolling deploy by design of Cloud.gov. Delete only the apps and routes, leave all services as we do not yet provision those automatically.
* Assign the label `Deploy with CircleCI` to this PR
* Observe the deploy-dev job output on CircleCI and confirm that the deployment is successful
* Using the [Cloud Foundry CLI](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html) extract the updated JWT_CERT:
```
cf login --sso
cf env tdp-backend | sed -n -e '/JWT_CERT/,/END CERTIFICATE/ p'
```
* Login to the [Login.gov Sandbox](https://dashboard.int.identitysandbox.gov/) and update the certificate there with the output from the step above. ([see also](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Technical-Documentation/jwt-key-rotation.md#devstaging-environments))
* Navigate to the deployed frontend application at `https://tdp-frontend.apps.cloud.gov` and confirm that the site works as intended.

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] Circle CI deploys to Cloud.gov using the new buildpack scripts
  - [x] Deployment arguments are dynamically pulled in as environment variables from the Circle CI host.
As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

N/A CI changes only

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A No UX changes

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*
